src/mon/OSDMonitor.cc

```diff
@@ -1663,6 +1665,12 @@ bool OSDMonitor::check_failure(utime_t now, int target_osd, failure_info_t& fi)
     for (map<int,failure_reporter_t>::iterator p = fi.reporters.begin();
 	 p != fi.reporters.end();
 	 ++p) {
+      //get subtree reporter
+      map<string,string> reporter_loc = osdmap.crush->get_full_location(p->first);
+      int osd_subtree_reporter  = osdmap.crush->get_item_id(reporter_loc[reporter_subtree_level]);
+      assert(osd_subtree_reporter != 0);
+      subtree_reporters.insert(osd_subtree_reporter);
```

Comment:
I don't think this is sufficient: we should make this a count, and when a reporter is canceled decrement the count.

I would put the subtree name (reporter_loc[reporter_subtree_level] inside the failure_reporter_t so that we don't recalculate it later.  And make that calculation when we get the report (prepare_failure).  Also please adjust the dout in that method to say the subtree level and the value that we use.. e.g.

<pre>
  dout(1) << "prepare_failure " << m->get_target()
	  << " from " << m->get_orig_source_inst()
	  << " (" << subtree_level << " " << subtree_value << ")"
          << " is reporting failure:" << m->if_osd_failed() << dendl;
</pre>

and make sure you adjust the by-subtree count in the cancel case in that function...